massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1057 @@
+    [
+        (Defined, Advance) => [Defined, Started, Failed],
+        (Started, Advance) => [Started, LockedIn, Failed],
+        (LockedIn, Advance) => [LockedIn, Active],
+        (Active, Advance) => Active,
+        (Failed, Advance) => Failed
+    ]
+);
+
+impl Defined {
+    /// Update state from state Defined
+    pub fn on_advance(self, input: Advance) -> MipState {
+        match input.now {
+            n if n > input.timeout => MipState::failed(),
+            n if n > input.start_timestamp => MipState::started(Amount::zero()),
+            _ => MipState::Defined(Defined {}),
+        }
+    }
+}
+
+impl Started {
+    /// Update state from state Started
+    pub fn on_advance(self, input: Advance) -> MipState {
+        if input.now > input.timeout {
+            return MipState::failed();
+        }
+
+        if input.threshold >= VERSIONING_THRESHOLD_TRANSITION_ACCEPTED {
+            MipState::locked_in()
+        } else {
+            MipState::started(input.threshold)
+        }
+    }
+}
+
+impl LockedIn {
+    /// Update state from state LockedIn ...
+    pub fn on_advance(self, input: Advance) -> MipState {
+        if input.now > input.timeout {
+            MipState::active()
+        } else {
+            MipState::locked_in()
+        }
+    }
+}
+
+impl Active {
+    /// Update state (will always stay in state Active)
+    pub fn on_advance(self, _input: Advance) -> Active {
+        Active {}
+    }
+}
+
+impl Failed {
+    /// Update state (will always stay in state Failed)
+    pub fn on_advance(self, _input: Advance) -> Failed {
+        Failed {}
+    }
+}
+
+/// Wrapper of MipState (in order to keep state history)
+#[derive(Debug, Clone, PartialEq)]
+pub struct MipStateHistory {
+    pub(crate) state: MipState,
+    history: BTreeMap<Advance, MipStateTypeId>,
+}
+
+impl MipStateHistory {
+    /// Create
+    pub fn new(defined: MassaTime) -> Self {
+        let state: MipState = Default::default();
+        let state_id = MipStateTypeId::from(&state);
+        // let mut advance = Advance::default();
+        // advance.now = defined;
+        let advance = Advance {
+            now: defined,
+            ..Default::default()
+        };
+        let history = BTreeMap::from([(advance, state_id)]);
+        Self {
+            state: Default::default(),
+            history,
+        }
+    }
+
+    /// Advance state
+    pub fn on_advance(&mut self, input: &Advance) {
+        let now = input.now;
+        // Check that input.now is not after last item in history
+        // We don't want to go backward
+        let is_forward = self
+            .history
+            .last_key_value()
+            .map(|(adv, _)| adv.now < now)
+            .unwrap_or(false);
+
+        if is_forward {
+            // machines crate (for state machine) does not support passing ref :-/
+            let state = self.state.on_advance(input.clone());
+            // Update history as well
+            if state != self.state {
```

Comment:
Here, is there a reason to update the history if we go from Started {threshold: 2} to Started {threshold: 3}?
Or should we consider it as duplicate as long as the threshold is not above the const?
This could add a lot of states to the history which is not optimal.
Another option, only keep the first "Started" and the latest one, removing all the states in between. The way we still have some info about "where we currently are"

Reply:
After discussion, this code will be changed to only keep the first Started state.

Reply:
fixed